mirgecom/eos.py

```diff
@@ -233,3 +232,236 @@ def total_energy(self, cv, pressure):
+
+    def kinetic_energy(self, cv: ConservedVars):
+        r"""Get kinetic (i.e. not internal) energy of gas.
+
+        The kinetic energy is calculated as:
+
+        .. math::
+
+            k = \frac{1}{2\rho}(\rho\vec{V} \cdot \rho\vec{V})
+        """
+        mom = cv.momentum
+        return (0.5 * np.dot(mom, mom) / cv.mass)
+
+    def internal_energy(self, cv: ConservedVars):
+        r"""Get internal thermal energy of gas.
+
+        The internal energy ($e$) is calculated as:
+
+        .. math::
+
+            e = \rho{E} - \frac{1}{2\rho}(\rho\vec{V} \cdot \rho\vec{V})
+        """
+        return (cv.energy - self.kinetic_energy(cv))
+
+    def get_density(self, pressure, temperature, species_fractions):
+        r"""Get the density from pressure, temperature, and species fractions (Y).
+
+        The gas density $\rho$ is calculated from pressure, temperature and $R$ as:
+
+        .. math::
+
+            \rho = \frac{p}{R_s T}
+        """
+        return self._pyrometheus_mech.get_density(pressure, temperature,
+                                                  species_fractions)
+
+    def get_internal_energy(self, temperature, species_fractions):
+        r"""Get the gas thermal energy from temperature, and species fractions (Y).
+
+        The gas internal energy $e$ is calculated from:
+
+        .. math::
+
+            e = R_s T \sum{Y_\alpha h_\alpha}
+        """
+        return self._pyrometheus_mech.get_mixture_internal_energy_mass(
+            temperature, species_fractions)
+
+    def get_species_molecular_weights(self):
+        """Get the species molecular weights."""
+        return self._pyrometheus_mech.wts
+
+    def get_production_rates(self, cv: ConservedVars):
+        """Get the production rate for each species."""
+        temperature = self.temperature(cv)
+        y = self.species_fractions(cv)
+        return self._pyrometheus_mech.get_net_production_rates(
+            cv.mass, temperature, y)
+
+    def species_fractions(self, cv: ConservedVars):
+        r"""Get species fractions $Y_\alpha$ from species mass density."""
+        return cv.species_mass / cv.mass
+
+    def pressure(self, cv: ConservedVars):
+        r"""Get thermodynamic pressure of the gas.
+
+        Gas pressure ($p$) is calculated from the internal energy ($e$) as:
+
+        .. math::
+
+            p = (\gamma_{\mathtt{mix}} - 1)e
+        """
+        temperature = self.temperature(cv)
+        y = self.species_fractions(cv)
+        return self._pyrometheus_mech.get_pressure(cv.mass, temperature, y)
+
+    def sound_speed(self, cv: ConservedVars):
+        r"""Get the speed of sound in the gas.
+
+        The speed of sound ($c$) is calculated as:
+
+        .. math::
+
+            c = \sqrt{\frac{\gamma_{\mathtt{mix}}{p}}{\rho}}
+        """
+        actx = cv.mass.array_context
+        c2 = (self.gamma(cv) * self.pressure(cv)) / cv.mass
+        return actx.np.sqrt(c2)
+
+    def temperature(self, cv: ConservedVars):
```

Comment:
Quite possibly, we should also permit a temperature guess to be supplied here, clarifying that we fall back to the one from the constructor if not provided.

Reply:
I agree.  This is potentially a sweeping change, and is not applicable to all EOS.  I'd like to be careful about aligning this (and other required changes) regarding the temperature handling with a strategy that we collectively agree on.  

Started an issue for this discussion: (#310)